# Guided Project: Designing and Creating a Database

In [94]:
import pandas as pd
import sqlite3
import csv
pd.set_option('max_columns', 180)
pd.set_option('max_rows', 200000)
pd.set_option('max_colwidth', 5000)

In [95]:
game_log = pd.read_csv('game_log.csv', low_memory=False)
park_codes = pd.read_csv('park_codes.csv')
person_codes = pd.read_csv('person_codes.csv')
team_codes = pd.read_csv('team_codes.csv')

In [96]:
game_log.shape

(171907, 161)

In [97]:
game_log.head()

date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18710504               0         Thu    CL1      NaN              1    FW1   
1  18710505               0         Fri    BS1      NaN              1    WS3   
2  18710506               0         Sat    CL1      NaN              2    RC1   
3  18710508               0         Mon    CL1      NaN              3    CH1   
4  18710509               0         Tue    BS1      NaN              2    TRO   

  h_league  h_game_number  v_score  h_score  length_outs day_night completion  \
0      NaN              1        0        2         54.0         D        NaN   
1      NaN              1       20       18         54.0         D        NaN   
2      NaN              1       12        4         54.0         D        NaN   
3      NaN              1       12       14         54.0         D        NaN   
4      NaN              1        9        5         54.0         D        NaN   

  forefeit protest park_id  attendance  length_minutes v_line_score  \
0      NaN     NaN   FOR01       200.0           120.0    000000000   
1      NaN     NaN   WAS01      5000.0           145.0    107000435   
2      NaN     NaN   RCK01      1000.0           140.0    610020003   
3      NaN     NaN   CHI01      5000.0           150.0    101403111   
4      NaN     NaN   TRO01      3250.0           145.0    000002232   

  h_line_score  v_at_bats  v_hits  v_doubles  v_triples  v_homeruns  v_rbi  \
0    010010000       30.0     4.0        1.0        0.0         0.0    0.0   
1    640113030       41.0    13.0        1.0        2.0         0.0   13.0   
2    010020100       49.0    11.0        1.0        1.0         0.0    8.0   
3    077000000       46.0    15.0        2.0        1.0         2.0   10.0   
4    101003000       46.0    17.0        4.0        1.0         0.0    6.0   

   v_sacrifice_hits  v_sacrifice_flies  v_hit_by_pitch  v_walks  \
0               0.0                0.0             0.0      1.0   
1               0.0                0.0             0.0     18.0   
2               0.0                0.0             0.0      0.0   
3               0.0                0.0             0.0      0.0   
4               0.0                0.0             0.0      2.0   

   v_intentional_walks  v_strikeouts  v_stolen_bases  v_caught_stealing  \
0                  NaN           6.0             1.0                NaN   
1                  NaN           5.0             3.0                NaN   
2                  NaN           1.0             0.0                NaN   
3                  NaN           1.0             0.0                NaN   
4                  NaN           0.0             1.0                NaN   

   v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
0                    -1.0                           NaN             4.0   
1                    -1.0                           NaN            12.0   
2                    -1.0                           NaN            10.0   
3                    -1.0                           NaN             7.0   
4                    -1.0                           NaN            12.0   

   v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
0              1.0                       1.0                 1.0   
1              1.0                       6.0                 6.0   
2              1.0                       0.0                 0.0   
3              1.0                       6.0                 6.0   
4              1.0                       2.0                 2.0   

   v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  v_passed_balls  \
0             0.0      0.0       27.0        9.0       0.0             3.0   
1             1.0      0.0       27.0       13.0      10.0             1.0   
2             2.0      0.0       27.0       12.0       8.0             5.0   
3             0.0      0.0       27.0       15.0      11.0             6.0   
4             0.0      0.0       

In [98]:
game_log['day_night'].value_counts(dropna=False)

D      82724
N      57426
NaN    31757
Name: day_night, dtype: int64

In [99]:
!cat game_log_fields.txt

Field(s)  Meaning
    1     Date in the form "yyyymmdd"
    2     Number of game:
             "0" -- a single game
             "1" -- the first game of a double (or triple) header
                    including seperate admission doubleheaders
             "2" -- the second game of a double (or triple) header
                    including seperate admission doubleheaders
             "3" -- the third game of a triple-header
             "A" -- the first game of a double-header involving 3 teams
             "B" -- the second game of a double-header involving 3 teams
    3     Day of week  ("Sun","Mon","Tue","Wed","Thu","Fri","Sat")
  4-5     Visiting team and league
    6     Visiting team game number
          For this and the home team game number, ties are counted as
          games and suspended games are counted from the starting
          rather than the ending date.
  7-8     Home team and league
    9     Home team game number
10-11     Visiting and home team score (unquoted)


In [100]:
park_codes.head()

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                        aka       city state       start  \
0                                       NaN     Albany    NY  09/11/1880   
1                                       NaN    Altoona    PA  04/30/1884   
2             Edison Field; Anaheim Stadium    Anaheim    CA  04/19/1966   
3                                       NaN  Arlington    TX  04/21/1972   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX  04/11/1994   

          end league                                           notes  
0  05/30/1882     NL  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1  05/31/1884     UA                                             NaN  
2         NaN     AL                                             NaN  
3  10/03/1993     AL                                             NaN  
4         NaN     AL                                             NaN

In [101]:
person_codes.head()

id     last   first player_debut mgr_debut coach_debut ump_debut
0  aardd001  Aardsma   David   04/06/2004       NaN         NaN       NaN
1  aaroh101    Aaron    Hank   04/13/1954       NaN         NaN       NaN
2  aarot101    Aaron  Tommie   04/10/1962       NaN  04/06/1979       NaN
3  aased001     Aase     Don   07/26/1977       NaN         NaN       NaN
4  abada001     Abad    Andy   09/10/2001       NaN         NaN       NaN

In [102]:
team_codes.head()

team_id league  start   end       city         nickname franch_id  seq
0     ALT     UA   1884  1884    Altoona  Mountain Cities       ALT    1
1     ARI     NL   1998     0    Arizona     Diamondbacks       ARI    1
2     BFN     NL   1879  1885    Buffalo           Bisons       BFN    1
3     BFP     PL   1890  1890    Buffalo           Bisons       BFP    1
4     BL1    NaN   1872  1874  Baltimore         Canaries       BL1    1

The game_log file contains 161 descriptive columns for over 170,000 baseball games, including data such as names of umpires, names of winning pitchers, the defensive positions of each player in each game, among others.

-  The park_codes file contains information about ballparks where games were played, connected to the game_log through the 'park_id' column
-  The person_codes file contains information about all people (players, managers, umpires) found in the game_log file. The 'id' column of person_codes matches id's found in the game_log file wherever an id is used to indicate a person
-  The team_codes table contains information regarding all teams which have recorded games in the game_log table. Each team has a unique ID recorded in the 'team_id' column. This column corresponds to any column in the game_log table with team_id information.

## Defensive Positions
-  1: Pitcher
-  2: Catcher
-  3: First Baseman
-  4: Second Baseman
-  5: Third Baseman
-  6: Shortstop
-  7: Left Field
-  8: Center Field
-  9: Right Field

In [103]:
team_codes['league'].unique()

array(['UA', 'NL', 'PL', nan, 'AA', 'AL', 'FL'], dtype=object)

## Leagues
-  UA: Union Association, 1884
-  NL: National League, one of two current leagues
-  PL: Players' League, 19th century league, ended 1890
-  AA: American Association, 1882-1891
-  AL: American League, the other current league
-  FL: Federal League, 1913-1915

In [104]:
# Create helper functions
DB = 'mlb.db'

def run_query(q):
    with sqlite3.connect(DB) as conn:
        return pd.read_sql(q, conn)
    
def run_command(c):
    with sqlite3.connect(DB) as conn:
        conn.execute('PRAGMA foreign_keys = ON')
        conn.isolation_level = None
        conn.execute(c)
        
def show_tables():
    q = '''
    SELECT
        name
        type
    FROM sqlite_master
    WHERE type IN ("table", "view");
    '''
    return run_query(q)

In [105]:
with sqlite3.connect(DB) as conn:
    game_log.to_sql('game_log', conn, index=False, if_exists='replace')
    park_codes.to_sql('park_codes', conn, index=False, if_exists='replace')
    person_codes.to_sql('person_codes', conn, index=False, if_exists='replace')
    team_codes.to_sql('team_codes', conn, index=False, if_exists='replace')

In [106]:
show_tables()

type
0                park
1              league
2      team_apperance
3              person
4     appearance_type
5          appearance
6                team
7                game
8     team_appearance
9   person_appearance
10           game_log
11         park_codes
12       person_codes
13         team_codes

In [107]:
c1 = '''
ALTER TABLE game_log
ADD COLUMN game_id;
'''
run_command(c1)

In [108]:
c2 = '''
UPDATE game_log
SET game_id = h_name || date || number_of_game;
'''
run_command(c2)

In [109]:
q1 = '''SELECT game_id, h_name, date, number_of_game FROM game_log LIMIT 5'''
run_query(q1)

game_id h_name      date  number_of_game
0  FW1187105040    FW1  18710504               0
1  WS3187105050    WS3  18710505               0
2  RC1187105060    RC1  18710506               0
3  CH1187105080    CH1  18710508               0
4  TRO187105090    TRO  18710509               0

## Opportunities for Normalization
-  park_codes: start, end, and league can all be found elsewhere
-  person_codes: debut columns can be derived from game_log
-  team_codes: start and end can be derived from game_log

In [110]:
c2 = '''
CREATE TABLE IF NOT EXISTS person (
    person_id TEXT PRIMARY KEY,
    first_name TEXT,
    last_name TEXT
);
'''
run_command(c2)

In [111]:
qq = 'SELECT * FROM person LIMIT 10'
run_query(qq)

person_id first_name    last_name
0  aardd001      David      Aardsma
1  aaroh101       Hank        Aaron
2  aarot101     Tommie        Aaron
3  aased001        Don         Aase
4  abada001       Andy         Abad
5  abadf001   Fernando         Abad
6  abadj101       John       Abadie
7  abbae101         Ed  Abbaticchio
8  abbeb101       Bert        Abbey
9  abbec101    Charlie        Abbey

In [112]:
c3 = '''
INSERT OR IGNORE INTO person
SELECT id, first, last FROM person_codes;'''
run_command(c3)

In [113]:
qq = 'SELECT * FROM person LIMIT 5'
run_query(qq)

person_id first_name last_name
0  aardd001      David   Aardsma
1  aaroh101       Hank     Aaron
2  aarot101     Tommie     Aaron
3  aased001        Don      Aase
4  abada001       Andy      Abad

In [114]:
c4 = '''
CREATE TABLE IF NOT EXISTS park
(
park_id TEXT PRIMARY KEY,
name TEXT,
nickname TEXT,
city TEXT,
state TEXT,
notes TEXT
);
'''
run_command(c4)

In [115]:
c5 = '''
INSERT OR IGNORE INTO park
SELECT
    park_id,
    name,
    aka,
    city,
    state,
    notes
FROM park_codes'''

run_command(c5)

In [116]:
qq = 'SELECT * FROM park LIMIT 5'
run_query(qq)

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                   nickname       city state  \
0                                      None     Albany    NY   
1                                      None    Altoona    PA   
2             Edison Field; Anaheim Stadium    Anaheim    CA   
3                                      None  Arlington    TX   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX   

                                            notes  
0  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1                                            None  
2                                            None  
3                                            None  
4                                            None

In [117]:
c6 = '''
CREATE TABLE IF NOT EXISTS league
(
league_id TEXT PRIMARY KEY,
name TEXT
);
'''
run_command(c6)

In [118]:
c7 = '''
INSERT OR IGNORE INTO league
VALUES
    ('NL', 'National League'),
    ('AL', 'American League'),
    ('UA', 'Union Association'),
    ('PL', "Players' League"),
    ('AA', 'American Association'),
    ('FL', 'Federal League')
;'''
run_command(c7)

In [119]:
qq = 'SELECT * FROM league'
run_query(qq)

league_id                  name
0        NL       National League
1        AL       American League
2        FL        Federal League
3        UA     Union Association
4        PL       Players' League
5        AA  American Association

In [120]:
cc = "DROP TABLE IF EXISTS appearance_type"
run_command(cc)

c8 = '''
CREATE TABLE appearance_type
(
appearance_type_id TEXT PRIMARY KEY,
name TEXT,
category TEXT
);'''
run_command(c8)

In [121]:
appearance_type = pd.read_csv('appearance_type.csv')
appearance_type.head()

appearance_type_id      name category
0                 O1  Batter 1  offense
1                 O2  Batter 2  offense
2                 O3  Batter 3  offense
3                 O4  Batter 4  offense
4                 O5  Batter 5  offense

In [122]:
BD = 'mlb.db'
with sqlite3.connect(BD) as conn:
    appearance_type.to_sql('appearance', conn, index=False, if_exists='replace')

In [123]:
c9 = '''
INSERT OR IGNORE INTO appearance_type
SELECT 
    appearance_type_id,
    name,
    category
FROM appearance
;'''
run_command(c9)

In [124]:
qq = 'SELECT * FROM appearance_type'
run_query(qq)

appearance_type_id                name category
0                  O1            Batter 1  offense
1                  O2            Batter 2  offense
2                  O3            Batter 3  offense
3                  O4            Batter 4  offense
4                  O5            Batter 5  offense
5                  O6            Batter 6  offense
6                  O7            Batter 7  offense
7                  O8            Batter 8  offense
8                  O9            Batter 9  offense
9                  D1             Pitcher  defense
10                 D2             Catcher  defense
11                 D3            1st Base  defense
12                 D4            2nd Base  defense
13                 D5            3rd Base  defense
14                 D6           Shortstop  defense
15                 D7          Left Field  defense
16                 D8        Center Field  defense
17                 D9         Right Field  defense
18                D10    Unknown Position  defense
19                UHP          Home Plate   umpire
20                U1B          First Base   umpire
21                U2B         Second Base   umpire
22                U3B          Third Base   umpire
23                ULF          Left Field   umpire
24                URF         Right Field   umpire
25                 MM             Manager  manager
26                AWP     Winning Pitcher    award
27                ALP      Losing Pitcher    award
28                ASP      Saving Pitcher    award
29                AWB  Winning RBI Batter    award
30                PSP    Starting Pitcher  pitcher

In [125]:
c10 = '''
CREATE TABLE IF NOT EXISTS team
(
team_id TEXT PRIMARY KEY,
league_id TEXT,
city TEXT,
nickname TEXT,
franch_id TEXT,
FOREIGN KEY (league_id) REFERENCES league(league_id)
);'''
run_command(c10)

In [126]:
c11 = '''
INSERT OR IGNORE INTO team
SELECT
    team_id,
    league,
    city,
    nickname,
    franch_id
FROM team_codes
;'''
run_command(c11)

In [127]:
qq = 'SELECT * FROM team LIMIT 5'
run_query(qq)

team_id league_id       city         nickname franch_id
0     ALT        UA    Altoona  Mountain Cities       ALT
1     ARI        NL    Arizona     Diamondbacks       ARI
2     BFN        NL    Buffalo           Bisons       BFN
3     BFP        PL    Buffalo           Bisons       BFP
4     BL1      None  Baltimore         Canaries       BL1

In [128]:
c12 = '''
CREATE TABLE IF NOT EXISTS game
(
game_id TEXT PRIMARY KEY,
date INTEGER,
number_of_game INTEGER,
park_id TEXT,
length_outs REAL,
day INTEGER,
completion TEXT,
forefeit TEXT,
protest TEXT,
attendance REAL,
length_minutes REAL,
additional_info TEXT,
acquisition_info TEXT,
FOREIGN KEY (park_id) REFERENCES park(park_id)
);
'''
run_command(c12)

In [129]:
c13 = '''
INSERT OR IGNORE INTO game
SELECT
    game_id,
    date,
    number_of_game,
    park_id,
    length_outs,
    day_night,
    completion,
    forefeit,
    protest,
    attendance,
    length_minutes,
    additional_info,
    acquisition_info
FROM game_log
;'''
run_command(c13)

In [130]:
qq = 'SELECT * FROM game LIMIT 10'
run_query(qq)

game_id      date  number_of_game park_id  length_outs day completion  \
0  FW1187105040  18710504               0   FOR01         54.0   D       None   
1  WS3187105050  18710505               0   WAS01         54.0   D       None   
2  RC1187105060  18710506               0   RCK01         54.0   D       None   
3  CH1187105080  18710508               0   CHI01         54.0   D       None   
4  TRO187105090  18710509               0   TRO01         54.0   D       None   
5  CL1187105110  18710511               0   CLE01         48.0   D       None   
6  CL1187105130  18710513               0   CIN01         54.0   D       None   
7  FW1187105130  18710513               0   FOR01         54.0   D       None   
8  FW1187105150  18710515               0   FOR01         54.0   D       None   
9  BS1187105160  18710516               0   BOS01         54.0   D       None   

  forefeit protest  attendance  length_minutes additional_info  \
0     None    None       200.0           120.0            None   
1     None    None      5000.0           145.0            HTBF   
2     None    None      1000.0           140.0            None   
3     None    None      5000.0           150.0            None   
4     None    None      3250.0           145.0            HTBF   
5        V    None      2500.0           120.0            None   
6     None    None      1200.0           150.0            None   
7     None    None      1500.0           105.0            None   
8     None    None         NaN           140.0            None   
9     None    None      2500.0             NaN            HTBF   

  acquisition_info  
0                Y  
1                Y  
2                Y  
3                Y  
4                Y  
5                Y  
6                Y  
7                Y  
8                Y  
9                Y

In [131]:
q5 = 'SELECT sql FROM sqlite_master WHERE name = "game_log" AND type = "table";'
run_query(q5)

sql
0  CREATE TABLE "game_log" (\n"date" INTEGER,\n  "number_of_game" INTEGER,\n  "day_of_week" TEXT,\n  "v_name" TEXT,\n  "v_league" TEXT,\n  "v_game_number" INTEGER,\n  "h_name" TEXT,\n  "h_league" TEXT,\n  "h_game_number" INTEGER,\n  "v_score" INTEGER,\n  "h_score" INTEGER,\n  "length_outs" REAL,\n  "day_night" TEXT,\n  "completion" TEXT,\n  "forefeit" TEXT,\n  "protest" TEXT,\n  "park_id" TEXT,\n  "attendance" REAL,\n  "length_minutes" REAL,\n  "v_line_score" TEXT,\n  "h_line_score" TEXT,\n  "v_at_bats" REAL,\n  "v_hits" REAL,\n  "v_doubles" REAL,\n  "v_triples" REAL,\n  "v_homeruns" REAL,\n  "v_rbi" REAL,\n  "v_sacrifice_hits" REAL,\n  "v_sacrifice_flies" REAL,\n  "v_hit_by_pitch" REAL,\n  "v_walks" REAL,\n  "v_intentional_walks" REAL,\n  "v_strikeouts" REAL,\n  "v_stolen_bases" REAL,\n  "v_caught_stealing" REAL,\n  "v_grounded_into_double" REAL,\n  "v_first_catcher_interference" REAL,\n  "v_left_on_base" REAL,\n  "v_pitchers_used" REAL,\n  "v_individual_earned_runs" REAL,\n  "v_team_earned_runs" REAL,\n  "v_wild_pitches" REAL,\n  "v_balks" REAL,\n  "v_putouts" REAL,\n  "v_assists" REAL,\n  "v_errors" REAL,\n  "v_passed_balls" REAL,\n  "v_double_plays" REAL,\n  "v_triple_plays" REAL,\n  "h_at_bats" REAL,\n  "h_hits" REAL,\n  "h_doubles" REAL,\n  "h_triples" REAL,\n  "h_homeruns" REAL,\n  "h_rbi" REAL,\n  "h_sacrifice_hits" REAL,\n  "h_sacrifice_flies" REAL,\n  "h_hit_by_pitch" REAL,\n  "h_walks" REAL,\n  "h_intentional_walks" REAL,\n  "h_strikeouts" REAL,\n  "h_stolen_bases" REAL,\n  "h_caught_stealing" REAL,\n  "h_grounded_into_double" REAL,\n  "h_first_catcher_interference" REAL,\n  "h_left_on_base" REAL,\n  "h_pitchers_used" REAL,\n  "h_individual_earned_runs" REAL,\n  "h_team_earned_runs" REAL,\n  "h_wild_pitches" REAL,\n  "h_balks" REAL,\n  "h_putouts" REAL,\n  "h_assists" REAL,\n  "h_errors" REAL,\n  "h_passed_balls" REAL,\n  "h_double_plays" REAL,\n  "h_triple_plays" REAL,\n  "hp_umpire_id" TEXT,\n  "hp_umpire_name" TEXT,\n  "1b_umpire_id" TEXT,\n  "1b_umpire_name" TEXT,\n  "2b_umpire_id" TEXT,\n  "2b_umpire_name" TEXT,\n  "3b_umpire_id" TEXT,\n  "3b_umpire_name" TEXT,\n  "lf_umpire_id" TEXT,\n  "lf_umpire_name" TEXT,\n  "rf_umpire_id" TEXT,\n  "rf_umpire_name" TEXT,\n  "v_manager_id" TEXT,\n  "v_manager_name" TEXT,\n  "h_manager_id" TEXT,\n  "h_manager_name" TEXT,\n  "winning_pitcher_id" TEXT,\n  "winning_pitcher_name" TEXT,\n  "losing_pitcher_id" TEXT,\n  "losing_pitcher_name" TEXT,\n  "saving_pitcher_id" TEXT,\n  "saving_pitcher_name" TEXT,\n  "winning_rbi_batter_id" TEXT,\n  "winning_rbi_batter_id_name" TEXT,\n  "v_starting_pitcher_id" TEXT,\n  "v_starting_pitcher_name" TEXT,\n  "h_starting_pitcher_id" TEXT,\n  "h_starting_pitcher_name" TEXT,\n  "v_player_1_id" TEXT,\n  "v_player_1_name" TEXT,\n  "v_player_1_def_pos" REAL,\n  "v_player_2_id" TEXT,\n  "v_player_2_name" TEXT,\n  "v_player_2_def_pos" REAL,\n  "v_player_3_id" TEXT,\n  "v_player_3_name" TEXT,\n  "v_player_3_def_pos" REAL,\n  "v_player_4_id" TEXT,\n  "v_player_4_name" TEXT,\n  "v_player_4_def_pos" REAL,\n  "v_player_5_id" TEXT,\n  "v_player_5_name" TEXT,\n  "v_player_5_def_pos" REAL,\n  "v_player_6_id" TEXT,\n  "v_player_6_name" TEXT,\n  "v_player_6_def_pos" REAL,\n  "v_player_7_id" TEXT,\n  "v_player_7_name" TEXT,\n  "v_player_7_def_pos" REAL,\n  "v_player_8_id" TEXT,\n  "v_player_8_name" TEXT,\n  "v_player_8_def_pos" REAL,\n  "v_player_9_id" TEXT,\n  "v_player_9_name" TEXT,\n  "v_player_9_def_pos" REAL,\n  "h_player_1_id" TEXT,\n  "h_player_1_name" TEXT,\n  "h_player_1_def_pos" REAL,\n  "h_player_2_id" TEXT,\n  "h_player_2_name" TEXT,\n  "h_player_2_def_pos" REAL,\n  "h_player_3_id" TEXT,\n  "h_player_3_name" TEXT,\n  "h_player_3_def_pos" REAL,\n  "h_player_4_id" TEXT,\n  "h_player_4_name" TEXT,\n  "h_player_4_def_pos" REAL,\n  "h_player_5_id" TEXT,\n  "h_player_5_name" TEXT,\n  "h_player_5_def_pos" REAL,\n  "h_player_6_id" TEXT,\n  "h_player_6_name" TEXT,\n  "h_player_6_def_pos" REAL,\n  "h_player_7_id" TEXT,\n  "h_player_7_name" TEXT,\n  "h_playe

In [132]:
c14 = '''
CREATE TABLE IF NOT EXISTS team_appearance
(
team_id TEXT,
game_id TEXT,
home INTEGER,
league_id TEXT,
score INTEGER,
line_score TEXT,
at_bats REAL,
hits REAL,
doubles REAL,
triples REAL,
homeruns REAL,
rbi REAL,
sacrifice_hits REAL,
sacrifice_flies REAL,
hit_by_pitch REAL,
walks REAL,
intentional_walks REAL,
strikeouts REAL,
stolen_bases REAL,
caught_stealing REAL,
grounded_into_double REAL,
first_catcher_interference REAL,
left_on_base REAL,
pitchers_used REAL,
individual_earned_runs REAL,
team_earned_runs REAL,
wild_pitches REAL,
balks REAL,
putouts REAL,
assists REAL,
errors REAL,
passed_balls REAL,
double_plays REAL,
triple_plays REAL,
PRIMARY KEY (team_id, game_id),
FOREIGN KEY (league_id) REFERENCES league(league_id)
);'''
run_command(c14)

In [133]:
c15 = '''
INSERT OR IGNORE INTO team_appearance
    SELECT
        h_name,
        game_id,
        1 AS home,
        h_league,
        h_score,
        h_line_score,
        h_at_bats,
        h_hits,
        h_doubles,
        h_triples,
        h_homeruns,
        h_rbi,
        h_sacrifice_hits,
        h_sacrifice_flies,
        h_hit_by_pitch,
        h_walks,
        h_intentional_walks,
        h_strikeouts,
        h_stolen_bases,
        h_caught_stealing,
        h_grounded_into_double,
        h_first_catcher_interference,
        h_left_on_base,
        h_pitchers_used,
        h_individual_earned_runs,
        h_team_earned_runs,
        h_wild_pitches,
        h_balks,
        h_putouts,
        h_assists,
        h_errors,
        h_passed_balls,
        h_double_plays,
        h_triple_plays
    FROM game_log

UNION

    SELECT
        v_name,
        game_id,
        0 AS home,
        v_league,
        v_score,
        v_line_score,
        v_at_bats,
        v_hits,
        v_doubles,
        v_triples,
        v_homeruns,
        v_rbi,
        v_sacrifice_hits,
        v_sacrifice_flies,
        v_hit_by_pitch,
        v_walks,
        v_intentional_walks,
        v_strikeouts,
        v_stolen_bases,
        v_caught_stealing,
        v_grounded_into_double,
        v_first_catcher_interference,
        v_left_on_base,
        v_pitchers_used,
        v_individual_earned_runs,
        v_team_earned_runs,
        v_wild_pitches,
        v_balks,
        v_putouts,
        v_assists,
        v_errors,
        v_passed_balls,
        v_double_plays,
        v_triple_plays
    FROM game_log;
'''
run_command(c15)

In [134]:
qq = 'SELECT * FROM team_appearance LIMIT 10'
run_query(qq)

team_id       game_id  home league_id  score line_score at_bats  hits  \
0     ALT  ALT188404300     1        UA      2       None    None  None   
1     ALT  ALT188405020     1        UA      3       None    None  None   
2     ALT  ALT188405030     1        UA      5       None    None  None   
3     ALT  ALT188405050     1        UA      2       None    None  None   
4     ALT  ALT188405100     1        UA      9       None    None  None   
5     ALT  ALT188405120     1        UA      3       None    None  None   
6     ALT  ALT188405140     1        UA      2       None    None  None   
7     ALT  ALT188405150     1        UA      7       None    None  None   
8     ALT  ALT188405160     1        UA      6       None    None  None   
9     ALT  ALT188405170     1        UA      8       None    None  None   

  doubles triples homeruns   rbi sacrifice_hits sacrifice_flies hit_by_pitch  \
0    None    None     None  None           None            None         None   
1    None    None     None  None           None            None         None   
2    None    None     None  None           None            None         None   
3    None    None     None  None           None            None         None   
4    None    None     None  None           None            None         None   
5    None    None     None  None           None            None         None   
6    None    None     None  None           None            None         None   
7    None    None     None  None           None            None         None   
8    None    None     None  None           None            None         None   
9    None    None     None  None           None            None         None   

  walks intentional_walks strikeouts stolen_bases caught_stealing  \
0  None              None       None         None            None   
1  None              None       None         None            None   
2  None              None       None         None            None   
3  None              None       None         None            None   
4  None              None       None         None            None   
5  None              None       None         None            None   
6  None              None       None         None            None   
7  None              None       None         None            None   
8  None              None       None         None            None   
9  None              None       None         None            None   

  grounded_into_double first_catcher_interference left_on_base pitchers_used  \
0                 None                       None         None          None   
1                 None                       None         None          None   
2                 None                       None         None          None   
3                 None                       None         None          None   
4                 None                       None         None          None   
5                 None                       None         None          None   
6                 None                       None         None          None   
7                 None                       None         None          None   
8                 None                       None         None          None   
9                 None                       None         None          None   

  individual_earned_runs team_earned_runs wild_pitches balks putouts assists  \
0                   None             None         None  None    None    None   
1                   None             None         None  None    None    None   
2                   None             None         None  None    None    None   
3                   None             None         None  None    None    None   
4                   None             None         None  None    None    None   
5                   None             None         None  None    None    None   
6                   None             None         None  None    None    None   
7             

In [135]:
qq = 'SELECT * FROM game_log WHERE game_id = "ALT188404300"'
run_query(qq)

date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18840430               0         Wed    SLU       UA              6    ALT   

  h_league  h_game_number  v_score  h_score length_outs day_night completion  \
0       UA              8       15        2        None      None       None   

  forefeit protest park_id attendance length_minutes v_line_score  \
0     None    None   ALT01       None           None         None   

  h_line_score v_at_bats v_hits v_doubles v_triples v_homeruns v_rbi  \
0         None      None   None      None      None       None  None   

  v_sacrifice_hits v_sacrifice_flies v_hit_by_pitch v_walks  \
0             None              None           None    None   

  v_intentional_walks v_strikeouts v_stolen_bases v_caught_stealing  \
0                None         None           None              None   

  v_grounded_into_double v_first_catcher_interference v_left_on_base  \
0                   None                         None           None   

  v_pitchers_used v_individual_earned_runs v_team_earned_runs v_wild_pitches  \
0            None                     None               None           None   

  v_balks v_putouts v_assists v_errors v_passed_balls v_double_plays  \
0    None      None      None     None           None           None   

  v_triple_plays h_at_bats h_hits h_doubles h_triples h_homeruns h_rbi  \
0           None      None   None      None      None       None  None   

  h_sacrifice_hits h_sacrifice_flies h_hit_by_pitch h_walks  \
0             None              None           None    None   

  h_intentional_walks h_strikeouts h_stolen_bases h_caught_stealing  \
0                None         None           None              None   

  h_grounded_into_double h_first_catcher_interference h_left_on_base  \
0                   None                         None           None   

  h_pitchers_used h_individual_earned_runs h_team_earned_runs h_wild_pitches  \
0            None                     None               None           None   

  h_balks h_putouts h_assists h_errors h_passed_balls h_double_plays  \
0    None      None      None     None           None           None   

  h_triple_plays hp_umpire_id    hp_umpire_name 1b_umpire_id 1b_umpire_name  \
0           None     maplb901  Blake Mapledoram         None           None   

  2b_umpire_id 2b_umpire_name 3b_umpire_id 3b_umpire_name lf_umpire_id  \
0         None           None         None           None         None   

  lf_umpire_name rf_umpire_id rf_umpire_name v_manager_id v_manager_name  \
0           None         None           None     sullt101   Ted Sullivan   

  h_manager_id h_manager_name winning_pitcher_id winning_pitcher_name  \
0     curte801      Ed Curtis               None                 None   

  losing_pitcher_id losing_pitcher_name saving_pitcher_id saving_pitcher_name  \
0              None                None              None                None   

  winning_rbi_batter_id winning_rbi_batter_id_name v_starting_pitcher_id  \
0                  None                       None              hodnc101   

  v_starting_pitcher_name h_starting_pitcher_id h_starting_pitcher_name  \
0                 Hodnett              murpj104                  Murphy   

  v_player_1_id v_player_1_name v_player_1_def_pos v_player_2_id  \
0          None            None               None          None   

  v_player_2_name v_player_2_def_pos v_player_3_id v_player_3_name  \
0            None               None          None            None   

  v_player_3_def_pos v_player_4_id v_player_4_name v_player_4_def_pos  \
0               None          None            None               None   

  v_player_5_id v_player_5_name v_player_5_def_pos v_player_6_id  \
0          None            None               None          None   

  v_player_6_name v_player_6_def_pos v_player_7_id v_player_7_name  \
0            None               None          None            None   

  v_player_7_def_pos v_player_

In [136]:
cc = "DROP TABLE IF EXISTS person_appearance"
run_command(cc)

c16 = '''
CREATE TABLE person_appearance
(
appearance_id INTEGER PRIMARY KEY,
person_id TEXT,
team_id TEXT,
game_id TEXT,
appearance_type_id TEXT,
FOREIGN KEY (person_id) REFERENCES person(person_id),
FOREIGN KEY (team_id) REFERENCES team(team_id),
FOREIGN KEY (game_id) REFERENCES game(game_id),
FOREIGN KEY (appearance_type_id) REFERENCES appearance_type(appearance_type_id)
);'''
run_command(c16)

In [137]:
!cat game_log_fields.txt

Field(s)  Meaning
    1     Date in the form "yyyymmdd"
    2     Number of game:
             "0" -- a single game
             "1" -- the first game of a double (or triple) header
                    including seperate admission doubleheaders
             "2" -- the second game of a double (or triple) header
                    including seperate admission doubleheaders
             "3" -- the third game of a triple-header
             "A" -- the first game of a double-header involving 3 teams
             "B" -- the second game of a double-header involving 3 teams
    3     Day of week  ("Sun","Mon","Tue","Wed","Thu","Fri","Sat")
  4-5     Visiting team and league
    6     Visiting team game number
          For this and the home team game number, ties are counted as
          games and suspended games are counted from the starting
          rather than the ending date.
  7-8     Home team and league
    9     Home team game number
10-11     Visiting and home team score (unquoted)


In [138]:
game_log.head()

date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18710504               0         Thu    CL1      NaN              1    FW1   
1  18710505               0         Fri    BS1      NaN              1    WS3   
2  18710506               0         Sat    CL1      NaN              2    RC1   
3  18710508               0         Mon    CL1      NaN              3    CH1   
4  18710509               0         Tue    BS1      NaN              2    TRO   

  h_league  h_game_number  v_score  h_score  length_outs day_night completion  \
0      NaN              1        0        2         54.0         D        NaN   
1      NaN              1       20       18         54.0         D        NaN   
2      NaN              1       12        4         54.0         D        NaN   
3      NaN              1       12       14         54.0         D        NaN   
4      NaN              1        9        5         54.0         D        NaN   

  forefeit protest park_id  attendance  length_minutes v_line_score  \
0      NaN     NaN   FOR01       200.0           120.0    000000000   
1      NaN     NaN   WAS01      5000.0           145.0    107000435   
2      NaN     NaN   RCK01      1000.0           140.0    610020003   
3      NaN     NaN   CHI01      5000.0           150.0    101403111   
4      NaN     NaN   TRO01      3250.0           145.0    000002232   

  h_line_score  v_at_bats  v_hits  v_doubles  v_triples  v_homeruns  v_rbi  \
0    010010000       30.0     4.0        1.0        0.0         0.0    0.0   
1    640113030       41.0    13.0        1.0        2.0         0.0   13.0   
2    010020100       49.0    11.0        1.0        1.0         0.0    8.0   
3    077000000       46.0    15.0        2.0        1.0         2.0   10.0   
4    101003000       46.0    17.0        4.0        1.0         0.0    6.0   

   v_sacrifice_hits  v_sacrifice_flies  v_hit_by_pitch  v_walks  \
0               0.0                0.0             0.0      1.0   
1               0.0                0.0             0.0     18.0   
2               0.0                0.0             0.0      0.0   
3               0.0                0.0             0.0      0.0   
4               0.0                0.0             0.0      2.0   

   v_intentional_walks  v_strikeouts  v_stolen_bases  v_caught_stealing  \
0                  NaN           6.0             1.0                NaN   
1                  NaN           5.0             3.0                NaN   
2                  NaN           1.0             0.0                NaN   
3                  NaN           1.0             0.0                NaN   
4                  NaN           0.0             1.0                NaN   

   v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
0                    -1.0                           NaN             4.0   
1                    -1.0                           NaN            12.0   
2                    -1.0                           NaN            10.0   
3                    -1.0                           NaN             7.0   
4                    -1.0                           NaN            12.0   

   v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
0              1.0                       1.0                 1.0   
1              1.0                       6.0                 6.0   
2              1.0                       0.0                 0.0   
3              1.0                       6.0                 6.0   
4              1.0                       2.0                 2.0   

   v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  v_passed_balls  \
0             0.0      0.0       27.0        9.0       0.0             3.0   
1             1.0      0.0       27.0       13.0      10.0             1.0   
2             2.0      0.0       27.0       12.0       8.0             5.0   
3             0.0      0.0       27.0       15.0      11.0             6.0   
4             0.0      0.0       

In [143]:
c17 = '''
INSERT OR IGNORE INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
)
    SELECT
        game_id,
        NULL,
        hp_umpire_id,
        'UHP'
    FROM game_log
    WHERE hp_umpire_id IS NOT NULL
    
UNION
    SELECT
        game_id,
        NULL,
        [1b_umpire_id],
        'U1B'
    FROM game_log
    WHERE [1b_umpire_id] IS NOT NULL
    
UNION

    SELECT
        game_id,
        NULL,
        [2b_umpire_id],
        'U2B'
    FROM game_log
    WHERE [2b_umpire_id] IS NOT NULL
    
UNION
    SELECT
        game_id,
        NULL,
        [3b_umpire_id],
        'U3B'
    FROM game_log
    WHERE [3b_umpire_id] IS NOT NULL
    
UNION

    SELECT
        game_id,
        NULL,
        lf_umpire_id,
        'ULF'
    FROM game_log
    WHERE lf_umpire_id IS NOT NULL
    
UNION

    SELECT
        game_id,
        NULL,
        rf_umpire_id,
        'URF'
    FROM game_log
    WHERE rf_umpire_id IS NOT NULL
    
UNION

    SELECT
        game_id,
        v_name,
        v_manager_id,
        'MM'
    FROM game_log
    WHERE v_manager_id IS NOT NULL
    
UNION

    SELECT
        game_id,
        h_name,
        h_manager_id,
        'MM'
    FROM game_log
    WHERE h_manager_id IS NOT NULL
    
UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        winning_pitcher_id,
        'AWP'
    FROM game_log
    WHERE winning_pitcher_id IS NOT NULL
    
UNION
    
    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN v_name
            ELSE h_name
            END,
        losing_pitcher_id,
        'ALP'
    FROM game_log
    WHERE losing_pitcher_id IS NOT NULL
    
UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        saving_pitcher_id,
        'ASP'
    FROM game_log
    WHERE saving_pitcher_id IS NOT NULL
    
UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        winning_rbi_batter_id,
        'AWB'
    FROM game_log
    WHERE winning_rbi_batter_id IS NOT NULL
    
UNION

    SELECT
        game_id,
        v_name,
        v_starting_pitcher_id,
        'PSP'
    FROM game_log
    WHERE v_starting_pitcher_id IS NOT NULL
    
UNION

    SELECT
        game_id,
        h_name,
        h_starting_pitcher_id,
        'PSP'
    FROM game_log
    WHERE h_starting_pitcher_id IS NOT NULL
;'''

run_command(c17)

In [144]:
c18 = """
INSERT INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
) 
    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "O{num}"
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL

UNION

    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "D" || CAST({hv}_player_{num}_def_pos AS INT)
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL;
"""
for hv in ["h", "v"]:
    for num in range(1,10):
        query_vars = {
            "hv": hv,
            "num": num
        }
run_command(c18.format(**query_vars))

In [152]:
print(run_query('SELECT COUNT(DISTINCT(game_id)) FROM person_appearance'))
print(run_query('SELECT COUNT(DISTINCT(game_id)) FROM game_log'))

   COUNT(DISTINCT(game_id))
0                    171907
   COUNT(DISTINCT(game_id))
0                    171907


In [153]:
show_tables()

type
0                park
1              league
2      team_apperance
3              person
4                team
5                game
6     team_appearance
7            game_log
8          park_codes
9        person_codes
10         team_codes
11    appearance_type
12         appearance
13  person_appearance

In [154]:
tables = [
    'game_log',
    'park_codes',
    'team_codes',
    'person_codes',
    'team_apperance'
]

for t in tables:
    c = 'DROP TABLE {}'.format(t)
    run_command(c)

show_tables()

type
0               park
1             league
2             person
3               team
4               game
5    team_appearance
6    appearance_type
7         appearance
8  person_appearance

Data has now been normalized into the 8 tables show above and the orginal non-normalized tables have been removed. 